### ISSUE TITLE
We should have a topfiles checker in tox

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8336]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8336|
|Type|enhancement|
|Created|2016-05-18 04:53:22Z|
|Branch|https://github.com/twisted/twisted/tree/topfile-checker-8336|

So then we can help catch the instances where a topfile is forgotten, and wrap it into our CI.

<details><summary>Searchable metadata</summary>

```
trac-id__8336 8336
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__normal normal
milestone__None None
branch__topfile_checker_8336 topfile-checker-8336
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1463547202991969 1463547202991969
changetime__1464556514664166 1464556514664166
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

Happy with this; please review. Please see the new buildbot builder.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban commented|
|-|-|

Is it ok to create multiple scripts for different checkers ?

I was thinking that maybe we can start a new tx-checker which has no false positives
and can be called on local branches without support for buildbot

We can make this a core principle of the new checker.

The current twistedchecker is more like a buildbot-twisted-checker

We can include the pyflakes checks... with the exception files and with support for running only files which were changed since trunk.

and we can include other checks for which we don't have false positives.

--------

I think that checking for topfiles is a good start, but we can start this tools with the goal of a generic tool.


Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @hawkowl|
|-|-|

Thanks for doing this, hawkie:

1. Please put the code in a module (similar to `twisted.python.release`) and write a few unit tests and docstrings for it.  Even for stuff like this, no cheating on code quality :).
2. In the future, when adding a new builder, please make it unsupported first, rather than making it supported and assuming stuff will land.

Fix point 1 and resubmit.

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl removed owner|
|-|-|

Okay, looking good now, with proper code coverage. Please review :)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Changes look good.

Happy to see this script.

Maybe we should have it documented somewhere, so that developers can run the script on their local branches before a commit.

I feel that using tox, with a separate virtualenv just for this is an overkill :)

--------

Here are my comments:


1. `CheckTopfileScript(print).main(sys.argv[1:])` looks strange to me. 

Maybe 

```
try:
    TopfileChecker(reported).main(args=sys.argv[1:])
except TopfileCheckerError as error:
    print(error.message)
    sys.exit(1)`
```

------

2. I have a similar checker for my project, but on the releese branch I check the reverse... the checker should fail if there are topfiles in the release branch :)

--------

I would say that 2 is worth having

Thanks again

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban commented|
|-|-|

There is one more comment.

Maybe is a good idea to check that we not only have a topfile, but the topfile has the same trac ID as the branch trac ID.

What do you think?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl set status to `closed`|
|-|-|

In changeset 86ef77d731d79b41f8325857c4f6943389b076f6
```
#!CommitTicketReference repository="" revision="86ef77d731d79b41f8325857c4f6943389b076f6"
Merge topfile-checker-8336: Add a topfile checker tox environment

Author: hawkowl
Reviewers: glyph, adiroiban
Fixes: #8336
```
